controllers/metal3.io/bmceventsubscription_controller.go

```diff
@@ -0,0 +1,278 @@
+/*
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	metal3v1alpha1 "github.com/metal3-io/baremetal-operator/apis/metal3.io/v1alpha1"
+	"github.com/metal3-io/baremetal-operator/pkg/ironic/bmc"
+	"github.com/metal3-io/baremetal-operator/pkg/provisioner"
+	"github.com/metal3-io/baremetal-operator/pkg/utils"
+)
+
+type BMCEventSubscriptionReconciler struct {
+	client.Client
+	Log                logr.Logger
+	ProvisionerFactory provisioner.Factory
+}
+
+func (r *BMCEventSubscriptionReconciler) Reconcile(ctx context.Context, request ctrl.Request) (result ctrl.Result, err error) {
+	reconcileSubscriptionCounters.With(subscriptionMetricLabels(request)).Inc()
+	defer func() {
+		if err != nil {
+			reconcileSubscriptionErrorCounter.Inc()
+		}
+	}()
+
+	reqLogger := r.Log.WithValues("bmceventsubscription", request.NamespacedName)
+	reqLogger.Info("start")
+
+	// Fetch the BMCEventSubscription
+	subscription := &metal3v1alpha1.BMCEventSubscription{}
+	err = r.Get(ctx, request.NamespacedName, subscription)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after
+			// reconcile request.  Owned objects are automatically
+			// garbage collected. For additional cleanup logic use
+			// finalizers.  Return and don't requeue
+			return ctrl.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return ctrl.Result{}, errors.Wrap(err, "could not load subscription")
+	}
+
+	if subscription.Spec.HostRef == "" {
+		return ctrl.Result{}, errors.New("Missing HostRef")
```

Comment:
This looks like a critical problem (a webhook could be perfect for that), and it could be useful to show it to user - otherwise there wouldn't be no clue about it if not looking in the logs. Given that it could be something that user may fix, I'd suggest to use re-evaluate the request after a while. Note that returning an error will produce an immediate re-enqueue (that will get rate-limited).

Reply:
I added a webhook.